pkg/analysis/markerscope/analyzer.go

```diff
@@ -0,0 +1,492 @@
+		}
+	}
+
+	return result
+}
+
+// defaultConfig applies default values to the configuration.
+func defaultConfig(cfg *MarkerScopeConfig) {
+	// Set default policy if not specified
+	if cfg.Policy == "" {
+		cfg.Policy = MarkerScopePolicyWarn
+	}
+}
+
+func (a *analyzer) run(pass *analysis.Pass) (any, error) {
+	inspector, ok := pass.ResultOf[inspectorhelper.Analyzer].(inspectorhelper.Inspector)
+	if !ok {
+		return nil, kalerrors.ErrCouldNotGetInspector
+	}
+
+	// Check field markers
+	inspector.InspectFields(func(field *ast.Field, _ extractjsontags.FieldTagInfo, markersAccess markershelper.Markers, _ string) {
+		a.checkFieldMarkers(pass, field, markersAccess)
+	})
+
+	// Check type markers
+	inspector.InspectTypeSpec(func(typeSpec *ast.TypeSpec, markersAccess markershelper.Markers) {
+		a.checkTypeSpecMarkers(pass, typeSpec, markersAccess)
+	})
+
+	return nil, nil //nolint:nilnil
+}
+
+// sortMarkersByPosition sorts markers by their position to ensure consistent ordering.
+func sortMarkersByPosition(markers []markershelper.Marker) []markershelper.Marker {
+	slices.SortFunc(markers, func(a, b markershelper.Marker) int {
+		if a.Pos < b.Pos {
+			return -1
+		}
+
+		if a.Pos > b.Pos {
+			return 1
+		}
+
+		return 0
+	})
+
+	return markers
+}
+
+// checkFieldMarkers checks markers on fields for violations.
+func (a *analyzer) checkFieldMarkers(pass *analysis.Pass, field *ast.Field, markersAccess markershelper.Markers) {
+	fieldMarkers := markersAccess.FieldMarkers(field)
+	markers := sortMarkersByPosition(fieldMarkers.UnsortedList())
+
+	for _, marker := range markers {
+		rule, ok := a.markerRules[marker.Identifier]
+		if !ok {
+			// No rule defined for this marker, skip validation
+			continue
+		}
+
+		// Check if FieldScope is allowed
+		if !rule.AllowsScope(FieldScope) {
+			a.reportFieldScopeViolation(pass, field, marker, rule)
+			continue
+		}
+
+		// Check type constraints if present
+		a.checkFieldTypeConstraintViolation(pass, field, marker, rule)
+	}
+}
+
+// checkTypeSpecMarkers checks markers on a type spec for violations.
+func (a *analyzer) checkTypeSpecMarkers(pass *analysis.Pass, typeSpec *ast.TypeSpec, markersAccess markershelper.Markers) {
+	typeMarkers := markersAccess.TypeMarkers(typeSpec)
+	markers := sortMarkersByPosition(typeMarkers.UnsortedList())
+
+	for _, marker := range markers {
+		rule, ok := a.markerRules[marker.Identifier]
+		if !ok {
+			// No rule defined for this marker, skip validation
+			continue
+		}
+
+		// Check if TypeScope is allowed
+		if !rule.AllowsScope(TypeScope) {
+			a.reportTypeScopeViolation(pass, typeSpec, marker, rule)
+			continue
+		}
+
+		// Check type constraints if present
+		a.checkTypeConstraintViolation(pass, typeSpec, marker, rule)
+	}
+}
+
+// reportFieldScopeViolation reports a scope violation for a field marker.
+func (a *analyzer) reportFieldScopeViolation(pass *analysis.Pass, field *ast.Field, marker markershelper.Marker, rule MarkerScopeRule) {
```

Comment:
What if instead of reporting and doing logic in here, we had a function roughly like this that could be called by both the field and type side of things

```
checkAllowedScope(pass *analysis.Pass, node ast.Node, marker markershelper.Marker, appliedScope ScopeConstraint, rule MarkerScopeRule) {
  if rule.AllowsScope(appliedScope) {
    // Do nothing, the applied scope is allowed
    return
  }

  // Do reporting and suggesting fix work
  message = fmt.Sprintf("marker %s cannot be applied to scope %s, must be applied to scope(s): %s", marker.Identifier, appliedScope, strings.Join(rule.Scopes, ","))
}
```